src/engine/sched.c

```diff
@@ -1086,16 +1145,75 @@ policy_fifo_enqueue(struct dss_xstream *dx, struct sched_request *req,
 		    void *prio_data)
 {
 	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_req_attr	*attr = &req->sr_attr;
+	int			 rc;
 
+	D_ASSERT(attr->sra_type < SCHED_REQ_TYPE_MAX);
+	/*
+	 * If @sra_enqueue_ts is not zero, this is a resent RPC, it will
+	 * be inserted to sorted heap. if it failed to insert because of
+	 * out of memory, use fifo list instead.
+	 */
+	if (attr->sra_enqueue_ts) {
+		rc = d_binheap_insert(&info->si_heap, &req->sr_node);
+		if (rc)
+			goto out;
+		return;
+	}
+out:
 	d_list_add_tail(&req->sr_link, &info->si_fifo_list);
 }
 
 static void
 policy_fifo_process(struct dss_xstream *dx)
 {
 	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_request	*req, *tmp, *req1;
+	struct d_binheap_node	*node;
+	d_list_t                 tmp_list;
+	int			 rc;
+
+	D_INIT_LIST_HEAD(&tmp_list);
+	/*
+	 * All retried RPCs are inserted into a sorted heap, they are sorted
+	 * by RPC enqueue time in the server side(firstly enqueue time). So
+	 * retried RPCs won't starve forever.
+	 */
+	d_list_for_each_entry_safe(req, tmp, &info->si_fifo_list, sr_link) {
+		while (!d_binheap_is_empty(&info->si_heap)) {
+			node = d_binheap_root(&info->si_heap);
+			req1 = container_of(node, struct sched_request, sr_node);
+			if (req1->sr_attr.sra_enqueue_ts < req->sr_attr.sra_enqueue_ts) {
+				rc = process_req(dx, req1);
+				if (rc > 0) {
+					d_binheap_remove(&info->si_heap, &req1->sr_node);
+					d_list_add_tail(&req->sr_link, &tmp_list);
+				}
+			} else {
+				break;
+			}
+		}
+		process_req(dx, req);
+	}
+
+	/* Process retried RPCs if any */
+	while (!d_binheap_is_empty(&info->si_heap)) {
+		node = d_binheap_root(&info->si_heap);
+		req1 = container_of(node, struct sched_request, sr_node);
+		rc = process_req(dx, req1);
+		if (rc > 0) {
+			d_binheap_remove(&info->si_heap, &req1->sr_node);
+			d_list_add_tail(&req->sr_link, &tmp_list);
+		}
+	}
 
-	process_req_list(dx, &info->si_fifo_list, false);
+	/*
+	 * Insert skipped retried RPCs back to heap.
+	 */
+	d_list_for_each_entry_safe(req, tmp, &tmp_list, sr_link) {
+		d_binheap_insert(&info->si_heap, &req->sr_node);
+		d_list_del_init(&req->sr_link);
+	}
```

Comment:
This remove & re-insert heap operation looks weird and inefficient. I'm wondering if we'd use this opportunity to track the requests per pool?